### PR TITLE
Task to move 3 of  hk2 modules from hk2 to Glassfish

### PR DESCRIPTION
**Pull request by sumauppa**
Moving hk2-config, config-generator plugin, config-types and ctm-exmple modules to Glassfish code. This has been placed under nucleus/hk2.

### CODE REVIEW

**Comment by sumauppa** [comment]
@glassfishrobot Run CI tests please

File: nucleus/pom.xml [hunk 1]
```diff
@@ -138,8 +138,8 @@
         <javax.el.version>3.0.1-b08</javax.el.version>
         <javax.el-api.version>3.0.1-b04</javax.el-api.version>
         <saaj-api.version>1.4.0</saaj-api.version>
-        <hk2.version>2.5.0-b44</hk2.version>
-        <hk2.plugin.version>2.5.0-b44</hk2.plugin.version>
+        <hk2.version>2.5.0-b57</hk2.version>
```
Comment (yaminikb) [inline nucleus/pom.xml]: The latest release of HK2 is 2.5.0-b60, you should be using that.

File: nucleus/pom.xml [hunk 2]
```diff
@@ -138,8 +138,8 @@
...
+        <hk2.plugin.version>2.5.0-b54</hk2.plugin.version>
```
Comment (yaminikb) [inline nucleus/pom.xml]: Please clarify why we are using 2 versions of HK2

File: nucleus/core/kernel/src/test/java/com/sun/enterprise/v3/server/AppServerStartupTest.java [hunk 3]
```diff
@@ -254,6 +255,7 @@ public void afterTest() {
      * the run level services are constructed and destroyed at the proper
      * run levels.
      */
+    @Ignore
```
Comment (yaminikb) [inline nucleus/core/kernel/src/test/java/com/sun/enterprise/v3/server/AppServerStartupTest.java]: 1. Add proper comments for why these tests are commented. 
2. Raise issue in GlassFish or HK2 for these failures.

Comment (sumauppa) [reply nucleus/core/kernel/src/test/java/com/sun/enterprise/v3/server/AppServerStartupTest.java]: These tests are failing after moving HK2 modules from HK2 to Glassfish. I will raise an issue once the changes are committed.  Will add a comment.

File: nucleus/pom.xml (hunk 1 above)
Comment (sumauppa) [reply nucleus/pom.xml]: I am using the version after removing HK2 modules from HK2 code. May be we can create an issue and update the hk2 versions. What do you think?

File: nucleus/pom.xml (hunk 2 above)
Comment (sumauppa) [reply nucleus/pom.xml]: If there are any updates in HK2 code and not in plugins, they only release hk2 and plugins will retain the same old hk2 version. So, we have different versions for hk2 and hk2 plugins.

**Comment by sumauppa** [comment]
CI tests are success, except 11 findbugs warnings in hk2-config module.

**Comment by sumauppa** [comment]
@glassfishrobot Run CI tests please

**Comment by romain-grecourt** [comment]
1. Why do we need to migrate ctm-example ?
With this change, the old HK2 config framework will be buried inside the GlassFish workspace, I don't think it needs example. If it we were keeping, I would have it under hk2/config-examples/ctm/

2. Unless Maven supports it with recent versions, it is not possible to build plugin and use as part of the same reactor (i.e same Maven invocation) ; this is sometimes possible but these are tricky glitches... You should consider releasing this plugin separately and make it generic so that it can use the right hk2-config from the build.

3. Why are you duplicating plugin configuration in all these pom.xml files ?

**Comment by sumauppa** [comment]
1)ctm-example is basically dependent on hk2-config and config-generator. So, we had to move this example to Glassfish. 
2)Initially had discussed with Yamini and then we thought to go ahead with this approach. Will Discuss again with her.
3)As we had to build the plugin in the same reactor, had to remove config-generator dependency from nucleus parent and placed it in required modules. Otherwise, we run into cycllic dependecies under nucleus/hk2.

**Comment by sumauppa** [comment]
@glassfishrobot Run CI tests please

**Comment by sumauppa** [comment]
1) Changed the direcory name from examples to config-examples (nucleus/hk2/config-examples). Ran CI tests after that.
2) As this is only used in Glassfish, we decided to keep in the same reactor. We made sure that this plugin is built first and ran all CTS and CI tests to make sure nothing is broken.
3)As plugin is in the same reactor, there would be cyclic dependency. Hence, removed that plugin from nucleus pom and placed it under the modues, wherever it is required.

**Review by jwells131313 — DISMISSED** [review]
I'm really happy this is getting done!  Thanks very much for doing it

**Comment by sumauppa** [comment]
@glassfishrobot Run CI tests please

**Comment by sumauppa** [comment]
@glassfishrobot Run CI tests please

**Comment by sumauppa** [comment]
@glassfishrobot Run CI tests please

**Comment by sumauppa** [comment]
@glassfishrobot Run CI tests please

**Comment by sumauppa** [comment]
@glassfishrobot Run CI tests please

**Comment by sumauppa** [comment]
@glassfishrobot Run CI tests please

**Review by romain-grecourt — DISMISSED** [review]
Approved.
Is there any workarounds leftover ?

**Comment by sumauppa** [comment]
Yes. We use in nucleus/hk2/pom.xml, overriding config-generator.

**Comment by sumauppa** [comment]
@glassfishrobot Run CI tests please

**Comment by sumauppa** [comment]
Moved all properties to nucleus parent.

**Comment by sumauppa** [comment]
@glassfishrobot Run CI tests please

**Review by yaminikb — APPROVED** [review]
Looks good.

**Pull request merged**